# eBay Kleinanzeigen Used Car Data

I will be working with this dataset of used cards from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle.

The aim of this project is to clean the data and analyze the included used car listings.

In [137]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding = 'Latin-1')

In [138]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [139]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Most of the columns have non-null values, but columns like vehicleType            , model, notRepairedDamage, gearbox, and fuelType all need to be looked at and possibly cleaned.

Most of the columns are strings with 15 columns, and 5 columns are integers.

In [140]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [141]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

I cleaned up the columns to do 2 things:
- Reword some of the columns to make them more readable.
- Change the column names form camelcase to snakecase

In [142]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Some of the columns are mostly one value and are therefore good candidates to be dropped.

Number of pictures seems to only contain 0s, that does not provide any information.

The price and odometer columns are numerical values that are stored as text. 


In [143]:
autos = (autos
         .drop(["seller", 
                "offer_type",
                "nr_of_pictures"],
               axis=1))

autos["odometer"] = (autos["odometer"]
                    .str.replace("km","")
                    .str.replace(",","")
                    .astype("float")
                    )

autos.rename(columns={"odometer":"odometer_km"},
             inplace=True)
autos["price"] = (autos["price"]
                 .str.replace("$","")
                 .str.replace(",","")
                 .astype("float"))

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_PS  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [144]:
print(autos["odometer_km"].unique().shape)
print("\n")
print(autos["odometer_km"].describe())
print("\n")
print(autos["odometer_km"].value_counts())

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64


In [145]:
print(autos["price"].unique().shape)
print("\n")
print(autos["price"].describe())
print("\n")
print(autos["price"]
      .value_counts()
      .sort_index(ascending=True))

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
8.0              1
9.0              1
10.0             7
11.0             2
12.0             3
13.0             2
14.0             1
15.0             2
17.0             3
18.0             1
20.0             4
25.0             5
29.0             1
30.0             7
35.0             1
40.0             6
45.0             4
47.0             1
49.0             4
50.0            49
55.0             2
59.0             1
60.0             9
65.0             5
66.0             1
              ... 
151990.0         1
155000.0         1
163500.0         1
163991.0         1
169000.0         1
169999.0         1
175000.0         1
180000.0         1
190000.0         1
194000.0         1
19700

In [146]:
autos = autos[autos["price"].between(0,350000)]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49986 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          49986 non-null object
name                  49986 non-null object
price                 49986 non-null float64
abtest                49986 non-null object
vehicle_type          44894 non-null object
registration_year     49986 non-null int64
gearbox               47310 non-null object
power_PS              49986 non-null int64
model                 47233 non-null object
odometer_km           49986 non-null float64
registration_month    49986 non-null int64
fuel_type             45509 non-null object
brand                 49986 non-null object
unrepaired_damage     40163 non-null object
ad_created            49986 non-null object
postal_code           49986 non-null int64
last_seen             49986 non-null object
dtypes: float64(2), int64(4), object(11)
memory usage: 6.9+ MB


After fixing up the price and odometer columns, I also removed any outliers from the columns. The odometer did not have any values that were too out there, but the price column had some extremely high values. The jump from 350,000 to 999,990 seemed a bit too large, so I removed any values above 350,000 for the price.

In [147]:
(autos["date_crawled"]
 .value_counts(normalize=True,
               dropna=False)
.sort_index())

2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
2016-03-05 14:07:26    0.00002
2016-03-05 14:07:40    0.00002
2016-03-05 14:07:45    0.00002
2016-03-05 14:08:00    0.00004
2016-03-05 14:08:05    0.00004
2016-03-05 14:08:27    0.00002
2016-03-05 14:08:42    0.00002
2016-03-05 14:09:02    0.00004
2016-03-05 14:09:05    0.00002
2016-03-05 14:09:20    0.00002
2016-03-05 14:09:22    0.00002
2016-03-05 14:09:38    0.00002
2016-03-05 14:09:46    0.00002
2016-03-05 14:09:56    0.00002
2016-03-05 14:09:57    0.00002
2016-03-05 14:09:58    0.00004
2016-03-05 14:10:18    0.00002
2016-03-05 14:10:20    0.00002
2016-03-05 14:10:46    0.00002
2016-03-05 14:11:03    0.00002
2016-03-05 14:11:05    0.00002
2016-03-05 14:11:14    0.00002
2016-03-05 14:11:15    0.00002
2016-03-05 14:11:25    0.00002
2016-03-05 14:11:40    0.00002
                        ...   
2016-04-07 10:36:19    0.00002
2016-04-

The distribution is pretty spread out, but the time causes the data to be difficult to understand. 

The date range for the date_crawled is about a month long.

In [148]:
(autos["ad_created"]
 .value_counts(normalize=True,
               dropna=False)
.sort_index())

2015-06-11 00:00:00    0.000020
2015-08-10 00:00:00    0.000020
2015-09-09 00:00:00    0.000020
2015-11-10 00:00:00    0.000020
2015-12-05 00:00:00    0.000020
2015-12-30 00:00:00    0.000020
2016-01-03 00:00:00    0.000020
2016-01-07 00:00:00    0.000020
2016-01-10 00:00:00    0.000040
2016-01-13 00:00:00    0.000020
2016-01-14 00:00:00    0.000020
2016-01-16 00:00:00    0.000020
2016-01-22 00:00:00    0.000020
2016-01-27 00:00:00    0.000060
2016-01-29 00:00:00    0.000020
2016-02-01 00:00:00    0.000020
2016-02-02 00:00:00    0.000040
2016-02-05 00:00:00    0.000040
2016-02-07 00:00:00    0.000020
2016-02-08 00:00:00    0.000020
2016-02-09 00:00:00    0.000040
2016-02-11 00:00:00    0.000020
2016-02-12 00:00:00    0.000060
2016-02-14 00:00:00    0.000040
2016-02-16 00:00:00    0.000020
2016-02-17 00:00:00    0.000020
2016-02-18 00:00:00    0.000040
2016-02-19 00:00:00    0.000060
2016-02-20 00:00:00    0.000040
2016-02-21 00:00:00    0.000060
                         ...   
2016-03-

The date range for this data is much longer than earlier, spanning about 10 months, but the bulk of the data is more recent.

In [149]:
(autos["last_seen"]
 .value_counts(normalize=True,
               dropna=False)
.sort_index())

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
2016-03-05 15:28:10    0.00002
2016-03-05 15:41:30    0.00002
2016-03-05 15:45:43    0.00002
2016-03-05 15:47:38    0.00002
2016-03-05 15:47:44    0.00002
2016-03-05 16:45:57    0.00002
2016-03-05 16:47:28    0.00002
2016-03-05 17:15:45    0.00002
2016-03-05 17:16:12    0.00002
2016-03-05 17:16:14    0.00002
2016-03-05 17:16:23    0.00002
2016-03-05 17:17:02    0.00002
2016-03-05 17:39:19    0.00002
2016-03-05 17:40:14    0.00002
2016-03-05 17:44:50    0.00002
2016-03-05 17:44:54    0.00002
2016-03-05 17:46:01    0.00002
2016-03-05 18:17:58    0.00002
2016-03-05 18:47:14    0.00002
2016-03-05 18:50:38    0.00002
2016-03-05 19:15:08    0.00002
2016-03-05 19:15:20    0.00002
2016-03-05 19:15:42    0.00002
2016-03-05 19:16:36    0.00002
2016-03-05 19:17:17    0.00002
                        ...   
2016-04-07 14:58:09    0.00004
2016-04-

Similar to the date_crawled data, the distribution has a date range of about a month. The distribution is split by the times, causing the data to be hard to read.

In [150]:
(autos["registration_year"]
 .describe())

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Some of the data presents outliers, such as the min being 1000 and the max being 9999. The data will have to be investigated.

In [151]:
(autos["registration_year"]
 .value_counts(normalize=True)
.sort_index())

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000040
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000660
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054055
2002    0.050674
2003    0.054555
2004    0.054755
2005    0.060317
2006    0.054155
2007    0.046093
2008    0.044632
2009    0.041952
2010    0.031949
2011    0.032689
2012    0.026467
2013    0.016125
2014    0.013304
2015    0.007982
2016    0.026327
2017    0.029048
2018    0.009823
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

Anything outside of the 1910-2016 range should be removed. Either the date was before cars were available to be registered or the registration year is after the point in time that the data was collected.

In [152]:
autos = autos[autos["registration_year"]
              .between(1910,2016)]
(autos["registration_year"]
 .value_counts(normalize=True)
.sort_index())

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000687
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003770
1990    0.008226
1991    0.007414
1992    0.008122
1993    0.009268
1994    0.013745
1995    0.027324
1996    0.030073
1997    0.042236
1998    0.051087
1999    0.062438
2000    0.069852
2001    0.056273
2002    0.052753
2003    0.056794
2004    0.057002
2005    0.062792
2006    0.056377
2007    0.047984
2008    0.046464
2009    0.043673
2010    0.033260
2011    0.034030
2012    0.027553
2013    0.016786
2014    0.0138

The data without outliers does not change too heavily, but it should be much more accurate.

In [153]:
brands = autos["brand"].value_counts(normalize=True,dropna=False)
brands

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
citroen           0.013912
smart             0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

To explore the brands column I will aggregate the values that are at least 5% of the dataset. That will include all the values from Ford and up.

In [154]:
brand_mean_prices = {}
brands = brands.head(6)
for brand,dist in brands.items():
    brand_bool = autos["brand"] == brand
    brand_mean_prices[brand] = (autos.loc[brand_bool,"price"].mean())
brand_mean_prices

{'audi': 9093.65003615329,
 'bmw': 8102.536248343744,
 'ford': 3652.095223880597,
 'mercedes_benz': 8485.239571958942,
 'opel': 2876.716403542549,
 'volkswagen': 5231.081983308787}

Volkswagon, which was the most popular brand, averaged out to 5231.08 and was cheaper than the BMW, Mercedes, and Audi. Those 3 values were 3 of the next 4 most popular brands. Opel is the 3rd most popular brand and was the 3rd lowest average price of the brands tested. Ford was another brand that is less expensive and rounds out the top 6 brands.

In [155]:
brand_mean_odometer = {}
for brand,dist in brands.items():
    brand_bool = autos["brand"] == brand
    brand_mean_odometer[brand] = (autos.loc[brand_bool,"odometer_km"].mean())
brand_mean_odometer

{'audi': 129287.78018799711,
 'bmw': 132431.38368351315,
 'ford': 124068.65671641791,
 'mercedes_benz': 130856.0821139987,
 'opel': 129223.14208702349,
 'volkswagen': 128724.10407461954}

In [156]:
brand_prices = pd.Series(brand_mean_prices)
brand_odometer = pd.Series(brand_mean_odometer)
brand_means = pd.DataFrame(brand_prices
                           ,columns=["mean_prices"])
brand_means["mean_odometer_km"] = brand_odometer
brand_means

mean_prices  mean_odometer_km
audi           9093.650036     129287.780188
bmw            8102.536248     132431.383684
ford           3652.095224     124068.656716
mercedes_benz  8485.239572     130856.082114
opel           2876.716404     129223.142087
volkswagen     5231.081983     128724.104075

The mean prices for each brand seem to be brand related. The average mileage for each brand is similar enough that it should not drastically affect the prices.

## Converting Dates to Integers

In [157]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [158]:
date_columns = ["date_crawled", "ad_created", "last_seen"]

for col in date_columns:
    autos[col] = (autos[col]
                 .str.split(expand=True)[0]
                 .str.replace("-","")
                 .astype('int'))

In [160]:
autos[["date_crawled", "ad_created", "last_seen"]].head()

date_crawled  ad_created  last_seen
0      20160326    20160326   20160406
1      20160404    20160404   20160406
2      20160326    20160326   20160406
3      20160312    20160312   20160315
4      20160401    20160401   20160401

In [161]:
autos.dtypes

date_crawled            int64
name                   object
price                 float64
abtest                 object
vehicle_type           object
registration_year       int64
gearbox                object
power_PS                int64
model                  object
odometer_km           float64
registration_month      int64
fuel_type              object
brand                  object
unrepaired_damage      object
ad_created              int64
postal_code             int64
last_seen               int64
dtype: object

In [166]:
(autos["date_crawled"]
 .value_counts(normalize=True,
               dropna=False)
.sort_index())

20160305    0.025262
20160306    0.014079
20160307    0.036155
20160308    0.033510
20160309    0.033364
20160310    0.032198
20160311    0.032364
20160312    0.036696
20160313    0.015745
20160314    0.036405
20160315    0.034030
20160316    0.029386
20160317    0.031698
20160318    0.012975
20160319    0.034801
20160320    0.037967
20160321    0.037467
20160322    0.032781
20160323    0.032343
20160324    0.029261
20160325    0.031614
20160326    0.032302
20160327    0.030761
20160328    0.034613
20160329    0.034155
20160330    0.033760
20160331    0.031864
20160401    0.033926
20160402    0.035488
20160403    0.038883
20160404    0.036613
20160405    0.012996
20160406    0.003103
20160407    0.001437
Name: date_crawled, dtype: float64

By removing the times, and changing the datatype to integers, the data can be understood much better.

## Translating Categories to English

In [169]:
print(autos["vehicle_type"].unique())
print("\n")
print(autos["gearbox"].unique())
print("\n")
print(autos["unrepaired_damage"].unique())

['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


['manuell' 'automatik' nan]


['nein' nan 'ja']


In [170]:
types_dict = {'bus':'bus',
         'limousine':'limousine',
         'kleinwagen':'compact',
         'kombi':'stationwagon',
         'coupe':'coupe',
         'suv':'suv',
         'cabrio':'convertible',
         'andere':'other'}
gear_dict = {'manuell':'manual',
            'automatik':'automatic'}
unrepaired_dict = {'nein':'no',
                  'ja':'yes'}
autos["vehicle_type"] = autos["vehicle_type"].map(types_dict)
autos["gearbox"] = autos["gearbox"].map(gear_dict)
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(unrepaired_dict)
print(autos["vehicle_type"].unique())
print("\n")
print(autos["gearbox"].unique())
print("\n")
print(autos["unrepaired_damage"].unique())

['bus' 'limousine' 'compact' 'stationwagon' nan 'coupe' 'suv'
 'convertible' 'other']


['manual' 'automatic' nan]


['no' nan 'yes']


## Most Common Brand/Model Combinations

In [187]:
autos["brand_model"]= autos['brand'].map(str) + " " + autos['model'].map(str)

autos["brand_model"].value_counts()

volkswagen golf                  3815
bmw 3er                          2688
volkswagen polo                  1677
opel corsa                       1645
opel astra                       1388
volkswagen passat                1388
audi a4                          1265
bmw 5er                          1163
mercedes_benz c_klasse           1147
mercedes_benz e_klasse            981
audi a3                           838
audi a6                           814
ford focus                        776
ford fiesta                       739
volkswagen transporter            688
renault twingo                    636
peugeot 2_reihe                   611
opel vectra                       558
smart fortwo                      552
mercedes_benz a_klasse            543
bmw 1er                           525
sonstige_autos nan                523
ford mondeo                       487
renault clio                      484
mercedes_benz andere              443
volkswagen touran                 438
volkswagen n